File: Python/data_structures/07_phone_book.py
import sys

MAX_SIZE = 10000000

class PhoneBook(object):
    def __init__(self, max_size):
        self._phone_book = [None] * max_size

    def add(self, number, name):
        self._phone_book[number] = name

    def delete(self, number):
        self._phone_book[number] = None

    def find(self, number):
        return self._phone_book[number]


def read_queries(reader, nr_queries):
    return [tuple(next(reader).split()) for _ in range(nr_queries)]


def process_queries(queries):
    phone_book = PhoneBook(MAX_SIZE)
    result = []
    for query in queries:
        operation = query[0]
        number = int(query[1])
        if operation == "add":
            phone_book.add(number, query[2])
        elif operation == "del":
            phone_book.delete(number)
        else:
            res = phone_book.find(number)
            if res is None:
                result.append("not found")
            else:
                result.append(str(res))
    return result


def main():
    reader = sys.stdin
    nr_queries = int(next(reader))
    queries = read_queries(reader, nr_queries)
    result = process_queries(queries)
    for item in result:
        print(item)


if __name__ == "__main__":
    main()

File: Haskell/data_structures/07-PhoneBook.hs
import Control.Monad (replicateM)
import Control.Monad.ST
import Control.Monad.Primitive (PrimMonad, PrimState)
import qualified Data.Vector.Mutable as MV

data Query = Addition Int String | Deletion Int | Find Int


maxSize :: Int
maxSize = 10000000


readQueries :: Int -> IO [Query]
readQueries nrQueries = do
    lines' <- replicateM nrQueries getLine
    return $ map (detectQueryType . words) lines'
    where
        detectQueryType :: [String] -> Query
        detectQueryType ["add", number, name] = Addition (read number) name
        detectQueryType ["del", number] = Deletion (read number)
        detectQueryType ["find", number] = Find (read number)


processQuery :: PrimMonad m => MV.MVector (PrimState m) (Maybe String) -> [String] -> [Query] -> m [String]
processQuery phoneBook acc [] = return $ reverse acc
processQuery phoneBook acc (Addition number name : qss) = do
    MV.write phoneBook number $ Just name
    processQuery phoneBook acc qss
processQuery phoneBook acc (Deletion number : qss) = do
    MV.write phoneBook number Nothing
    processQuery phoneBook acc qss
processQuery phoneBook acc (Find number : qss) = do
    contact <- MV.read phoneBook number
    case contact of Nothing -> processQuery phoneBook ("not found" : acc) qss
                    Just name -> processQuery phoneBook (name : acc) qss



processQueries :: [Query] -> [String]
processQueries queries = runST $ do
    phoneBook <- MV.replicate maxSize (Nothing :: Maybe String)
    processQuery phoneBook [] queries


main :: IO()
main = do
    nrQueries <- readLn
    queries <- readQueries nrQueries
    let results = processQueries queries
    mapM_ putStrLn $ results
